File: src/compiler/parser/visitors/TokenTypedAdapterVisitor.java
package compiler.parser.visitors;

import compiler.lexer.token.*;

import static compiler.lexer.token.KeywordToken.*;
import static compiler.lexer.token.OperatorToken.*;
import static compiler.lexer.token.SymbolToken.*;

public interface TokenTypedAdapterVisitor<T> extends TokenVisitor<T> {
  @Override
  default T visit(CommentToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(EOFToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(WhitespaceToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(IdentifierToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(ProgramKeywordToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(MainKeywordToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(FunctionKeywordToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(ClassKeywordToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(TypeToken.FloatKeywordToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(TypeToken.IntegerKeywordToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(TypeToken.StringKeywordToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(IfKeywordToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(ElseIfKeywordToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(ElseKeywordToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(WhileKeywordToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(InputKeywordToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(PrintKeywordToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(NewKeywordToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(ReturnKeywordToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(VarKeywordToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(LessThan token) {
    return visitChildren(token);
  }

  @Override
  default T visit(GreaterThan token) {
    return visitChildren(token);
  }

  @Override
  default T visit(Asterisk token) {
    return visitChildren(token);
  }

  @Override
  default T visit(Equal token) {
    return visitChildren(token);
  }

  @Override
  default T visit(Minus token) {
    return visitChildren(token);
  }

  @Override
  default T visit(Plus token) {
    return visitChildren(token);
  }

  @Override
  default T visit(Ampersand token) {
    return visitChildren(token);
  }

  @Override
  default T visit(Arrow token) {
    return visitChildren(token);
  }

  @Override
  default T visit(EqualEqual token) {
    return visitChildren(token);
  }

  @Override
  default T visit(NotEqual token) {
    return visitChildren(token);
  }

  @Override
  default T visit(LessThanOrEqual token) {
    return visitChildren(token);
  }

  @Override
  default T visit(GreaterThanOrEqual token) {
    return visitChildren(token);
  }

  @Override
  default T visit(BitShiftLeft token) {
    return visitChildren(token);
  }

  @Override
  default T visit(BitShiftRight token) {
    return visitChildren(token);
  }

  @Override
  default T visit(FloatToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(IntegerToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(StringToken token) {
    return visitChildren(token);
  }

  @Override
  default T visit(Comma token) {
    return visitChildren(token);
  }

  @Override
  default T visit(SemiColon token) {
    return visitChildren(token);
  }

  @Override
  default T visit(LeftBrace token) {
    return visitChildren(token);
  }

  @Override
  default T visit(RightBrace token) {
    return visitChildren(token);
  }

  @Override
  default T visit(LeftBracket token) {
    return visitChildren(token);
  }

  @Override
  default T visit(RightBracket token) {
    return visitChildren(token);
  }

  @Override
  default T visit(LeftParen token) {
    return visitChildren(token);
  }

  @Override
  default T visit(RightParen token) {
    return visitChildren(token);
  }

  @Override
  default T visit(Caret token) {
    return visitChildren(token);
  }

  @Override
  default T visit(Colon token) {
    return visitChildren(token);
  }

  @Override
  default T visit(Period token) {
    return visitChildren(token);
  }

  @Override
  default T visit(ForwardSlash token) {
    return visitChildren(token);
  }

  default T visitChildren(Token token) {
    for (final var child : token.children) {
      ((Token) child).accept(this);
    }
    return null;
  }

}

File: src/compiler/parser/visitors/TokenTypedAdapterVisitor.kt
package compiler.parser.visitors

import compiler.parser.Symbol.Terminal
import compiler.parser.Symbol.Terminal.Ignorable.*
import compiler.parser.Symbol.Terminal.Keyword.*
import compiler.parser.Symbol.Terminal.Keyword.Type.*
import compiler.parser.Symbol.Terminal.Operator.*
import compiler.parser.Symbol.Terminal.Punctuation.*
import compiler.parser.Symbol.Terminal.TypedTerminal.*

interface TokenTypedAdapterVisitor<T> : TokenVisitor<T> {
    fun defaultValue(): T

    override fun visit(token: CommentTerminal): T {
        return visitChildren(token)
    }

    override fun visit(token: EOFTerminal): T {
        return visitChildren(token)
    }

    override fun visit(token: WhitespaceTerminal): T {
        return visitChildren(token)
    }

    override fun visit(token: IdentifierTerminal): T {
        return visitChildren(token)
    }

    override fun visit(token: ProgramKeyword): T {
        return visitChildren(token)
    }

    override fun visit(token: MainKeyword): T {
        return visitChildren(token)
    }

    override fun visit(token: FunctionKeyword): T {
        return visitChildren(token)
    }

    override fun visit(token: ClassKeyword): T {
        return visitChildren(token)
    }

    override fun visit(token: FloatKeyword): T {
        return visitChildren(token)
    }

    override fun visit(token: IntegerKeyword): T {
        return visitChildren(token)
    }

    override fun visit(token: StringKeyword): T {
        return visitChildren(token)
    }

    override fun visit(token: Void): T {
        return visitChildren(token)
    }

    override fun visit(token: IfKeyword): T {
        return visitChildren(token)
    }

    override fun visit(token: ElseIfKeyword): T {
        return visitChildren(token)
    }

    override fun visit(token: ElseKeyword): T {
        return visitChildren(token)
    }

    override fun visit(token: WhileKeyword): T {
        return visitChildren(token)
    }

    override fun visit(token: InputKeyword): T {
        return visitChildren(token)
    }

    override fun visit(token: PrintKeyword): T {
        return visitChildren(token)
    }

    override fun visit(token: NewKeyword): T {
        return visitChildren(token)
    }

    override fun visit(token: ReturnKeyword): T {
        return visitChildren(token)
    }

    override fun visit(token: VarKeyword): T {
        return visitChildren(token)
    }

    override fun visit(token: LessThan): T {
        return visitChildren(token)
    }

    override fun visit(token: GreaterThan): T {
        return visitChildren(token)
    }

    override fun visit(token: Asterisk): T {
        return visitChildren(token)
    }

    override fun visit(token: Equal): T {
        return visitChildren(token)
    }

    override fun visit(token: Minus): T {
        return visitChildren(token)
    }

    override fun visit(token: Plus): T {
        return visitChildren(token)
    }

    override fun visit(token: Ampersand): T {
        return visitChildren(token)
    }

    override fun visit(token: Arrow): T {
        return visitChildren(token)
    }

    override fun visit(token: EqualEqual): T {
        return visitChildren(token)
    }

    override fun visit(token: NotEqual): T {
        return visitChildren(token)
    }

    override fun visit(token: LessThanOrEqual): T {
        return visitChildren(token)
    }

    override fun visit(token: GreaterThanOrEqual): T {
        return visitChildren(token)
    }

    override fun visit(token: BitShiftLeft): T {
        return visitChildren(token)
    }

    override fun visit(token: BitShiftRight): T {
        return visitChildren(token)
    }

    override fun visit(token: FloatTerminal): T {
        return visitChildren(token)
    }

    override fun visit(token: IntegerTerminal): T {
        return visitChildren(token)
    }

    override fun visit(token: StringTerminal): T {
        return visitChildren(token)
    }

    override fun visit(token: Comma): T {
        return visitChildren(token)
    }

    override fun visit(token: SemiColon): T {
        return visitChildren(token)
    }

    override fun visit(token: LeftBrace): T {
        return visitChildren(token)
    }

    override fun visit(token: RightBrace): T {
        return visitChildren(token)
    }

    override fun visit(token: LeftBracket): T {
        return visitChildren(token)
    }

    override fun visit(token: RightBracket): T {
        return visitChildren(token)
    }

    override fun visit(token: LeftParen): T {
        return visitChildren(token)
    }

    override fun visit(token: RightParen): T {
        return visitChildren(token)
    }

    override fun visit(token: Caret): T {
        return visitChildren(token)
    }

    override fun visit(token: Colon): T {
        return visitChildren(token)
    }

    override fun visit(token: Period): T {
        return visitChildren(token)
    }

    override fun visit(token: ForwardSlash): T {
        return visitChildren(token)
    }

    fun visitChildren(terminal: Terminal): T {
        for (child in terminal.children) {
            this.accept(child as Terminal)
        }
        return defaultValue()
    }
}
